File: java_src/src/test/java/com/google/crypto/tink/hybrid/HpkeParametersTest.java
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.hybrid;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.assertThrows;

import java.security.GeneralSecurityException;
import org.junit.Test;
import org.junit.experimental.theories.DataPoints;
import org.junit.experimental.theories.FromDataPoints;
import org.junit.experimental.theories.Theories;
import org.junit.experimental.theories.Theory;
import org.junit.runner.RunWith;

@RunWith(Theories.class)
public final class HpkeParametersTest {

  @DataPoints("variants")
  public static final HpkeParameters.Variant[] VARIANTS =
      new HpkeParameters.Variant[] {
        HpkeParameters.Variant.TINK,
        HpkeParameters.Variant.CRUNCHY,
        HpkeParameters.Variant.NO_PREFIX,
      };

  @DataPoints("kemIds")
  public static final HpkeParameters.KemId[] KEM_IDS =
      new HpkeParameters.KemId[] {
        HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256,
        HpkeParameters.KemId.DHKEM_P384_HKDF_SHA384,
        HpkeParameters.KemId.DHKEM_P521_HKDF_SHA512,
        HpkeParameters.KemId.DHKEM_X25519_HKDF_SHA256,
      };

  @DataPoints("kdfIds")
  public static final HpkeParameters.KdfId[] KDF_IDS =
      new HpkeParameters.KdfId[] {
        HpkeParameters.KdfId.HKDF_SHA256,
        HpkeParameters.KdfId.HKDF_SHA384,
        HpkeParameters.KdfId.HKDF_SHA512,
      };

  @DataPoints("aeadIds")
  public static final HpkeParameters.AeadId[] AEAD_IDS =
      new HpkeParameters.AeadId[] {
        HpkeParameters.AeadId.AES_128_GCM,
        HpkeParameters.AeadId.AES_256_GCM,
        HpkeParameters.AeadId.CHACHA20_POLY1305,
      };

  @Theory
  public void buildParameters(
      @FromDataPoints("variants") HpkeParameters.Variant variant,
      @FromDataPoints("kemIds") HpkeParameters.KemId kemId,
      @FromDataPoints("kdfIds") HpkeParameters.KdfId kdfId,
      @FromDataPoints("aeadIds") HpkeParameters.AeadId aeadId)
      throws Exception {
    HpkeParameters params =
        HpkeParameters.builder()
            .setVariant(variant)
            .setKemId(kemId)
            .setKdfId(kdfId)
            .setAeadId(aeadId)
            .build();

    assertThat(params.getVariant()).isEqualTo(variant);
    assertThat(params.getKemId()).isEqualTo(kemId);
    assertThat(params.getKdfId()).isEqualTo(kdfId);
    assertThat(params.getAeadId()).isEqualTo(aeadId);
  }

  @Theory
  public void buildParametersWithDefaultVariant(
      @FromDataPoints("kemIds") HpkeParameters.KemId kemId,
      @FromDataPoints("kdfIds") HpkeParameters.KdfId kdfId,
      @FromDataPoints("aeadIds") HpkeParameters.AeadId aeadId)
      throws Exception {
    HpkeParameters params =
        HpkeParameters.builder().setKemId(kemId).setKdfId(kdfId).setAeadId(aeadId).build();

    assertThat(params.getVariant()).isEqualTo(HpkeParameters.Variant.NO_PREFIX);
    assertThat(params.getKemId()).isEqualTo(kemId);
    assertThat(params.getKdfId()).isEqualTo(kdfId);
    assertThat(params.getAeadId()).isEqualTo(aeadId);
  }

  @Test
  public void buildParameters_failsWithoutKemId() throws Exception {
    HpkeParameters.Builder builder =
        HpkeParameters.builder()
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM);

    assertThrows(GeneralSecurityException.class, builder::build);
  }

  @Test
  public void buildParameters_failsWithoutKdfId() throws Exception {
    HpkeParameters.Builder builder =
        HpkeParameters.builder()
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM);

    assertThrows(GeneralSecurityException.class, builder::build);
  }

  @Test
  public void buildParameters_failsWithoutAeadId() throws Exception {
    HpkeParameters.Builder builder =
        HpkeParameters.builder()
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256);

    assertThrows(GeneralSecurityException.class, builder::build);
  }

  @Test
  public void hasIdRequirement() throws Exception {
    HpkeParameters noPrefixParams =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.NO_PREFIX)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();
    assertThat(noPrefixParams.hasIdRequirement()).isFalse();

    HpkeParameters tinkParams =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.TINK)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();
    assertThat(tinkParams.hasIdRequirement()).isTrue();

    HpkeParameters crunchyParams =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.CRUNCHY)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();
    assertThat(crunchyParams.hasIdRequirement()).isTrue();
  }

  @Test
  public void sameParamsAreEqual() throws Exception {
    HpkeParameters params =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.NO_PREFIX)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();

    HpkeParameters duplicateParams =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.NO_PREFIX)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();

    assertThat(params).isEqualTo(duplicateParams);
    assertThat(params.hashCode()).isEqualTo(duplicateParams.hashCode());
  }

  @Test
  public void differentVariantsAreNotEqual() throws Exception {
    HpkeParameters params =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.NO_PREFIX)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();

    HpkeParameters differentVariant =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.TINK)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();

    assertThat(params).isNotEqualTo(differentVariant);
    assertThat(params.hashCode()).isNotEqualTo(differentVariant.hashCode());
  }

  @Test
  public void differentKemAreNotEqual() throws Exception {
    HpkeParameters params =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.NO_PREFIX)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();

    HpkeParameters differentKem =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.NO_PREFIX)
            .setKemId(HpkeParameters.KemId.DHKEM_P384_HKDF_SHA384)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();

    assertThat(params).isNotEqualTo(differentKem);
    assertThat(params.hashCode()).isNotEqualTo(differentKem.hashCode());
  }

  @Test
  public void differentKdfsAreNotEqual() throws Exception {
    HpkeParameters params =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.NO_PREFIX)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();

    HpkeParameters differentKdf =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.NO_PREFIX)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA384)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();

    assertThat(params).isNotEqualTo(differentKdf);
    assertThat(params.hashCode()).isNotEqualTo(differentKdf.hashCode());
  }

  @Test
  public void differentAeadsAreNotEqual() throws Exception {
    HpkeParameters params =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.NO_PREFIX)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_128_GCM)
            .build();

    HpkeParameters differentAead =
        HpkeParameters.builder()
            .setVariant(HpkeParameters.Variant.NO_PREFIX)
            .setKemId(HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256)
            .setKdfId(HpkeParameters.KdfId.HKDF_SHA256)
            .setAeadId(HpkeParameters.AeadId.AES_256_GCM)
            .build();

    assertThat(params).isNotEqualTo(differentAead);
    assertThat(params.hashCode()).isNotEqualTo(differentAead.hashCode());
  }
}

File: cc/hybrid/hpke_parameters_test.cc
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

#include "tink/hybrid/hpke_parameters.h"

#include <tuple>

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/status/status.h"
#include "tink/util/statusor.h"
#include "tink/util/test_matchers.h"

namespace crypto {
namespace tink {
namespace {

using ::crypto::tink::test::IsOk;
using ::crypto::tink::test::StatusIs;
using ::testing::Combine;
using ::testing::Eq;
using ::testing::IsTrue;
using ::testing::TestWithParam;
using ::testing::Values;

struct VariantWithIdRequirement {
  HpkeParameters::Variant variant;
  bool has_id_requirement;
};

using HpkeParametersTest =
    TestWithParam<std::tuple<HpkeParameters::KemId, HpkeParameters::KdfId,
                             HpkeParameters::AeadId, VariantWithIdRequirement>>;

INSTANTIATE_TEST_SUITE_P(
    HpkeParametersTestSuite, HpkeParametersTest,
    Combine(Values(HpkeParameters::KemId::kDhkemP256HkdfSha256,
                   HpkeParameters::KemId::kDhkemP384HkdfSha384,
                   HpkeParameters::KemId::kDhkemP521HkdfSha512,
                   HpkeParameters::KemId::kDhkemX25519HkdfSha256),
            Values(HpkeParameters::KdfId::kHkdfSha256,
                   HpkeParameters::KdfId::kHkdfSha384,
                   HpkeParameters::KdfId::kHkdfSha512),
            Values(HpkeParameters::AeadId::kAesGcm128,
                   HpkeParameters::AeadId::kAesGcm256,
                   HpkeParameters::AeadId::kChaCha20Poly1305),
            Values(VariantWithIdRequirement{HpkeParameters::Variant::kTink,
                                            /*has_id_requirement=*/true},
                   VariantWithIdRequirement{HpkeParameters::Variant::kCrunchy,
                                            /*has_id_requirement=*/true},
                   VariantWithIdRequirement{HpkeParameters::Variant::kNoPrefix,
                                            /*has_id_requirement=*/false})));

TEST_P(HpkeParametersTest, Build) {
  HpkeParameters::KemId kem_id;
  HpkeParameters::KdfId kdf_id;
  HpkeParameters::AeadId aead_id;
  VariantWithIdRequirement variant;
  std::tie(kem_id, kdf_id, aead_id, variant) = GetParam();

  util::StatusOr<HpkeParameters> parameters = HpkeParameters::Builder()
                                                  .SetVariant(variant.variant)
                                                  .SetKemId(kem_id)
                                                  .SetKdfId(kdf_id)
                                                  .SetAeadId(aead_id)
                                                  .Build();
  ASSERT_THAT(parameters, IsOk());

  EXPECT_THAT(parameters->GetKemId(), Eq(kem_id));
  EXPECT_THAT(parameters->GetKdfId(), Eq(kdf_id));
  EXPECT_THAT(parameters->GetAeadId(), Eq(aead_id));
  EXPECT_THAT(parameters->GetVariant(), Eq(variant.variant));
  EXPECT_THAT(parameters->HasIdRequirement(), Eq(variant.has_id_requirement));
}

TEST(HpkeParametersTest, BuildWithInvalidVariantFails) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::
                          kDoNotUseInsteadUseDefaultWhenWritingSwitchStatements)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  EXPECT_THAT(parameters.status(),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(HpkeParametersTest, BuildWithoutVariantFails) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  EXPECT_THAT(parameters.status(),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(HpkeParametersTest, BuildWithInvalidKemIdFails) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::
                        kDoNotUseInsteadUseDefaultWhenWritingSwitchStatements)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  EXPECT_THAT(parameters.status(),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(HpkeParametersTest, BuildWithoutKemIdFails) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  EXPECT_THAT(parameters.status(),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(HpkeParametersTest, BuildWithInvalidKdfIdFails) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::
                        kDoNotUseInsteadUseDefaultWhenWritingSwitchStatements)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  EXPECT_THAT(parameters.status(),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(HpkeParametersTest, BuildWithoutKdfIdFails) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  EXPECT_THAT(parameters.status(),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(HpkeParametersTest, BuildWithInvalidAeadIdFails) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::
                         kDoNotUseInsteadUseDefaultWhenWritingSwitchStatements)
          .Build();
  EXPECT_THAT(parameters.status(),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(HpkeParametersTest, BuildWithoutAeadIdFails) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .Build();
  EXPECT_THAT(parameters.status(),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(Ed25519ParametersTest, CopyConstructor) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  ASSERT_THAT(parameters, IsOk());

  HpkeParameters copy(*parameters);

  EXPECT_THAT(copy.GetVariant(), Eq(HpkeParameters::Variant::kTink));
  EXPECT_THAT(copy.HasIdRequirement(), IsTrue());
}

TEST(Ed25519ParametersTest, CopyAssignment) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  ASSERT_THAT(parameters, IsOk());

  HpkeParameters copy = *parameters;

  EXPECT_THAT(copy.GetVariant(), Eq(HpkeParameters::Variant::kTink));
  EXPECT_THAT(copy.HasIdRequirement(), IsTrue());
}

TEST_P(HpkeParametersTest, ParametersEquals) {
  HpkeParameters::KemId kem_id;
  HpkeParameters::KdfId kdf_id;
  HpkeParameters::AeadId aead_id;
  VariantWithIdRequirement variant;
  std::tie(kem_id, kdf_id, aead_id, variant) = GetParam();

  util::StatusOr<HpkeParameters> parameters = HpkeParameters::Builder()
                                                  .SetVariant(variant.variant)
                                                  .SetKemId(kem_id)
                                                  .SetKdfId(kdf_id)
                                                  .SetAeadId(aead_id)
                                                  .Build();
  ASSERT_THAT(parameters, IsOk());

  util::StatusOr<HpkeParameters> other_parameters =
      HpkeParameters::Builder()
          .SetVariant(variant.variant)
          .SetKemId(kem_id)
          .SetKdfId(kdf_id)
          .SetAeadId(aead_id)
          .Build();
  ASSERT_THAT(other_parameters, IsOk());

  EXPECT_TRUE(*parameters == *other_parameters);
  EXPECT_TRUE(*other_parameters == *parameters);
  EXPECT_FALSE(*parameters != *other_parameters);
  EXPECT_FALSE(*other_parameters != *parameters);
}

TEST(HpkeParametersTest, VariantNotEqual) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  ASSERT_THAT(parameters, IsOk());

  util::StatusOr<HpkeParameters> other_parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kCrunchy)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  ASSERT_THAT(other_parameters, IsOk());

  EXPECT_TRUE(*parameters != *other_parameters);
  EXPECT_FALSE(*parameters == *other_parameters);
}

TEST(HpkeParametersTest, KemIdNotEqual) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  ASSERT_THAT(parameters, IsOk());

  util::StatusOr<HpkeParameters> other_parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemP256HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  ASSERT_THAT(other_parameters, IsOk());

  EXPECT_TRUE(*parameters != *other_parameters);
  EXPECT_FALSE(*parameters == *other_parameters);
}

TEST(HpkeParametersTest, KdfIdNotEqual) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha384)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  ASSERT_THAT(parameters, IsOk());

  util::StatusOr<HpkeParameters> other_parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  ASSERT_THAT(other_parameters, IsOk());

  EXPECT_TRUE(*parameters != *other_parameters);
  EXPECT_FALSE(*parameters == *other_parameters);
}

TEST(HpkeParametersTest, AeadIdNotEqual) {
  util::StatusOr<HpkeParameters> parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kAesGcm128)
          .Build();
  ASSERT_THAT(parameters, IsOk());

  util::StatusOr<HpkeParameters> other_parameters =
      HpkeParameters::Builder()
          .SetVariant(HpkeParameters::Variant::kTink)
          .SetKemId(HpkeParameters::KemId::kDhkemX25519HkdfSha256)
          .SetKdfId(HpkeParameters::KdfId::kHkdfSha256)
          .SetAeadId(HpkeParameters::AeadId::kChaCha20Poly1305)
          .Build();
  ASSERT_THAT(other_parameters, IsOk());

  EXPECT_TRUE(*parameters != *other_parameters);
  EXPECT_FALSE(*parameters == *other_parameters);
}

}  // namespace
}  // namespace tink
}  // namespace crypto
